engine/src/cubos/engine/collisions/broad_phase/plugin.cpp

```diff
@@ -0,0 +1,199 @@
+                }
+
+                sweepAndPrune->activePerAxis[axis].insert(marker.entity);
+            }
+            else
+            {
+                sweepAndPrune->activePerAxis[axis].erase(marker.entity);
+            }
+        }
+    }
+}
+
+BroadPhaseCandidates::CollisionType getCollisionType(bool box, bool capsule)
+{
+    if (box && capsule)
+    {
+        return BroadPhaseCandidates::CollisionType::BoxCapsule;
+    }
+
+    if (box)
+    {
+        return BroadPhaseCandidates::CollisionType::BoxBox;
+    }
+
+    return BroadPhaseCandidates::CollisionType::CapsuleCapsule;
+}
+
+/// @brief Finds all pairs of colliders which may be colliding.
+///
+/// @details
+/// TODO: This query is disgusting. We need a way to find if a component is present without reading it.
+/// Maybe something like Commands but for reads?
+void findPairs(Query<OptRead<BoxCollisionShape>, OptRead<CapsuleCollisionShape>, Read<Collider>> query,
+               Read<BroadPhaseSweepAndPrune> sweepAndPrune, Write<BroadPhaseCandidates> candidates)
+{
+    candidates->clearCandidates();
+
+    for (glm::length_t axis = 0; axis < 3; axis++)
+    {
+        for (const auto& [entity, overlaps] : sweepAndPrune->sweepOverlapMaps[axis])
+        {
+            auto [box, capsule, collider] = query[entity].value();
+            for (const auto& other : overlaps)
+            {
+                auto [otherBox, otherCapsule, otherCollider] = query[other].value();
+
+                // TODO: Should this be inside the if statement?
+                auto type = getCollisionType(box || otherBox, capsule || otherCapsule);
+
+                switch (axis)
+                {
+                case 0: // X
+                    if (collider->worldAABB.overlapsY(otherCollider->worldAABB) &&
+                        collider->worldAABB.overlapsZ(otherCollider->worldAABB))
+                    {
+                        candidates->addCandidate(type, {entity, other});
+                    }
+                    break;
+                case 1: // Y
+                    if (collider->worldAABB.overlapsX(otherCollider->worldAABB) &&
+                        collider->worldAABB.overlapsZ(otherCollider->worldAABB))
+                    {
+                        candidates->addCandidate(type, {entity, other});
+                    }
+                    break;
+                case 2: // Z
+                    if (collider->worldAABB.overlapsX(otherCollider->worldAABB) &&
+                        collider->worldAABB.overlapsY(otherCollider->worldAABB))
+                    {
+                        candidates->addCandidate(type, {entity, other});
+                    }
+                    break;
+                }
+            }
+        }
+    }
+}
+
+void cubos::engine::broadPhaseCollisionsPlugin(Cubos& cubos)
+{
+    // FIXME: Is it ok not to add resources from the general plugin?
+
+    cubos.addResource<BroadPhaseCandidates>();
+    cubos.addResource<BroadPhaseSweepAndPrune>();
```

Comment:
Why not add them from the broad phase plugin?

Reply:
I want it to be explicit that public resources like the shapes and collider are added by the plugin. Adding those in broadphase would be confusing.

This was just a problem when broadphase was public, now it's fine not to add them here.